### ISSUE TITLE
Generalizing SoftMaxWithLossLayer

### ISSUE THREAD

**Issue by lolongcovas** [BVLC/caffe#2463]
Hi,

I've modified the loss function of softmax_loss_layer.cpp so that it generalizes the loss function in the case of imbalance and balanced data:

(Eq 1)
![codecogseqn](https://cloud.githubusercontent.com/assets/3662732/7650299/456c2098-faf5-11e4-8f21-7440c8445ef3.gif)

in the case of balance data, the previous loss function becomes:

(Eq2)
![codecogseqn 1](https://cloud.githubusercontent.com/assets/3662732/7650316/67d06234-faf5-11e4-96a2-e7b262d2d778.gif)

Alright, my code works fine (for balanced and imbalanced data). However, it doesnt pass the test case of ignoring specific label (at line 66 of test_softmax_with_loss_layer.cpp):

```
TYPED_TEST(SoftmaxWithLossLayerTest, TestForwardIgnoreLabel) {
  typedef typename TypeParam::Dtype Dtype;
  LayerParameter layer_param;
  layer_param.mutable_loss_param()->set_normalize(false);
  // First, compute the loss with all labels
  scoped_ptr<SoftmaxWithLossLayer<Dtype> > layer(
      new SoftmaxWithLossLayer<Dtype>(layer_param));
  layer->SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
  layer->Forward(this->blob_bottom_vec_, this->blob_top_vec_);
  Dtype full_loss = this->blob_top_loss_->cpu_data()[0];
  // Now, accumulate the loss, ignoring each label in {0, ..., 4} in turn.
  Dtype accum_loss = 0;
  for (int label = 0; label < 5; ++label) {
    layer_param.mutable_loss_param()->set_ignore_label(label);
    layer.reset(new SoftmaxWithLossLayer<Dtype>(layer_param));
    layer->SetUp(this->blob_bottom_vec_, this->blob_top_vec_);
    layer->Forward(this->blob_bottom_vec_, this->blob_top_vec_);
    accum_loss += this->blob_top_loss_->cpu_data()[0];
  }
  // Check that each label was included all but once.
  EXPECT_NEAR(4 * full_loss, accum_loss, 1e-4);
}
```

this test case works fine for (Eq2). But, in the case of (Eq1), the loss varies according to ```fyi```. Therefore, the ```EXPECT_NEAR(4 * full_loss, accum_loss, 1e-4);``` fails.

My question is: how can I modify this test case so that it tests the case of ignoring specific label when the data is balanced and imbalanced?

Best

**Comment by longjon** [comment]
This tracker is for issues and bugs with Caffe; please ask coding and usage questions on the [caffe-users](https://groups.google.com/forum/#!forum/caffe-users) list. Thanks!
